upload/catalog/controller/extension/payment/authorizenet_accept_hosted.php

```diff
@@ -0,0 +1,259 @@
+							'settingValue' => json_encode(array(
+								'show' => false,
+								'required' => false
+							))
+						),
+						array(
+							'settingName' => 'hostedPaymentIFrameCommunicatorUrl',
+							'settingValue' => json_encode(array(
+								'url' => $this->url->link('extension/payment/authorizenet_accept_hosted/callback')
+							))
+						)
+					)
+				)
+			)
+		));
+
+		$api_url = 'https://api.authorize.net/xml/v1/request.api';
+		if ($this->config->get('payment_authorizenet_accept_hosted_testmode')) $api_url = 'https://apitest.authorize.net/xml/v1/request.api';
+
+		$curl = curl_init($api_url);
+		curl_setopt($curl, CURLOPT_PORT, 443);
+		curl_setopt($curl, CURLOPT_HEADER, 0);
+		curl_setopt($curl, CURLOPT_SSL_VERIFYPEER, 1);
+		curl_setopt($curl, CURLOPT_RETURNTRANSFER, 1);
+		curl_setopt($curl, CURLOPT_FORBID_REUSE, 1);
+		curl_setopt($curl, CURLOPT_FRESH_CONNECT, 1);
+		curl_setopt($curl, CURLOPT_POST, 1);
+		curl_setopt($curl, CURLOPT_CONNECTTIMEOUT, 10);
+		curl_setopt($curl, CURLOPT_TIMEOUT, 10);
+		curl_setopt($curl, CURLOPT_POSTFIELDS, $request);
+
+		if ($this->config->get('payment_authorizenet_accept_hosted_debug')) $this->log->write('AUTHORIZE.NET ACCEPT HOSTED SEND: ' . $request);
+
+		$response_raw = curl_exec($curl);
+		if ($this->config->get('payment_authorizenet_accept_hosted_debug')) $this->log->write('AUTHORIZE.NET ACCEPT HOSTED RECV: ' . $response_raw);
+
+		if (curl_error($curl)) {
+			$data['error'] = $this->language->get('error_no_form');
+			$this->log->write('AUTHORIZE.NET ACCEPT HOSTED CURL ERROR: ' . curl_errno($curl) . '::' . curl_error($curl));
+		} elseif (!$response_raw) {
+			$this->log->write('AUTHORIZE.NET ACCEPT HOSTED CURL ERROR: No response');
+		} else {
+			if (substr($response_raw, 0, 3) === "\xef\xbb\xbf") {
+				$response_raw = substr($response_raw, 3); //result of RFC-violating inclusion of BOM by authorize.net; remove if they fix their bug
+			}
+
+			$response = json_decode($response_raw);
+			if ($response === null) $data['error'] = $this->language->get('error_no_form');
+			else {
+				if ($response->messages->resultCode != 'Ok') {
+					$data['error'] = $this->language->get('error_no_form');
+					$this->log->write('Authorize.net accept hosted returned error while getting payment form: ' .
+						$response->messages->resultCode . ': ' . $response->messages->message[0]->code . ' ' . $response->messages->message[0]->text);
+				}
+				else {
+					$data['token'] = $response->token;
+				}
+			}
+		}
+		$data['token_responder'] = 'https://accept.authorize.net/payment/payment';
+		if ($this->config->get('payment_authorizenet_accept_hosted_testmode')) $data['token_responder'] = 'https://test.authorize.net/payment/payment';
+
+		return $this->load->view('extension/payment/authorizenet_accept_hosted', $data);
+	}
+
+	public function callback() {
+		$data = array();
+		$this->response->setOutput($this->load->view('extension/payment/authorizenet_accept_hosted_callback', $data));
```

Comment:
I needed to insert at line 150: 

    header("Content-Security-Policy: frame-ancestors 'self' https://test.authorize.net");

Otherwise the browser denied loading of the callback URL by the offsite page, loaded in the `iframe` in file upload/catalog/view/theme/default/template/extension/payment/authorizenet_accept_hosted.twig, line 5. 

The "Pay" and "Cancel" buttons did not appear on the page until after adding the `Content-Security-Policy: frame-ancestors` header.